### DISCUSSION TITLE
Multithreading use tilecache

### DISCUSSION THREAD

**Discussion by shadowerzc** [ppiastucki/recast4j#141] [Q&A] [answered]
I need your some suggest. I need use main thread to update obstacle, then anther thread pools only findpath to improve speed. Becuse of concurrency problems, I will use CopyOnWriteArrayList replace ArrayList. Is that righe? I need your help.

Navmesh.class
before change:
private final Map<Integer, List<MeshTile>> posLookup = new HashMap<>();
after change: 
private final Map<Integer, CopyOnWriteArrayList<MeshTile>> posLookup = new HashMap<>();

And Will there be some other problems ?

**Comment by ppiastucki** [comment, marked as answer]
The library is not designed to handle navmesh updates and pathfinding concurrently. The place you mentioned could be made thread-safe but there might be other places to change too. Using CopyOnWriteArrayList will not work as it would also require replacing the reference to the list.
The proper approach would be to use a Lock or a Semaphore to make sure navmesh updates do no overlap with ongoing path searches.
E.g.
```
    ReentrantReadWriteLock rwLock = new ReentrantReadWriteLock(true);

    void updateNavmesh() {
        rwLock.writeLock().lock();
        try {
            // run navmesh update here
        } finally {
            rwLock.writeLock().unlock();
        }
    }
	
    void findPath() {
        rwLock.readLock().lock();
        try {
            // run findPath here
        } finally {
            rwLock.readLock().unlock();
        }
    }
```
